### PR TITLE
Adds "addme" method to serviced

### PR DESCRIPTION
**Pull request by ed-**
A delegate host can now add itself to a main host
with a new add-private / "addme" command, which also
registers the host using an agreed-upon keypair stored at
/opt/serviced/etc/common.key

### CODE REVIEW

File: cli/cmd/host.go [hunk 1]
```diff
@@ -286,6 +317,76 @@ func (c *ServicedCli) cmdHostAdd(ctx *cli.Context) {
 	c.outputDelegateKey(host, nat, privateKey, keyfileName, registerHost)
 }
 
+// serviced host add-private HOST:PORT POOLID [--memory SIZE|%] [--nat-address HOST:PORT]
+func (c *ServicedCli) cmdHostAddPrivate(ctx *cli.Context) {
```
Comment (jplouis) [inline cli/cmd/host.go]: For this method if there is an error, please exit with a non-zero code.

File: cli/api/host.go [hunk 2]
```diff
@@ -172,6 +172,49 @@ func (a *api) AddHost(config HostConfig) (*host.Host, []byte, error) {
 	}
 }
 
+// Adds a new host and uses a common key to register it.
+func (a *api) AddHostPrivate(config HostConfig) (*host.Host, []byte, error) {
```
Comment (jplouis) [inline cli/api/host.go]: Document the return values

File: cli/cmd/key.go [hunk 3]
```diff
@@ -169,3 +171,8 @@ func (c *ServicedCli) outputDelegateKey(host *host.Host, nat utils.URL, keyData
 	}
 	fmt.Println(host.ID)
 }
+
+func (c *ServicedCli) outputCommonKey(host *host.Host, nat utils.URL, keyData []byte) {
```
Comment (jplouis) [inline cli/cmd/key.go]: Document the method

File: facade/host.go [hunk 4]
```diff
@@ -61,6 +61,21 @@ func (f *Facade) AddHost(ctx datastore.Context, entity *host.Host) ([]byte, erro
 	return key, alog.Error(err)
 }
 
+// AddHost registers a host with serviced. Returns the host's _public_ key.
+// Returns an error if host already exists or if the host's IP is a virtual IP.
+func (f *Facade) AddHostPrivate(ctx datastore.Context, entity *host.Host) ([]byte, error) {
+	defer ctx.Metrics().Stop(ctx.Metrics().Start("Facade.AddHostPrivate"))
+	alog := f.auditLogger.Message(ctx, "Adding Host").Action(audit.Add).Entity(entity)
```
Comment (jplouis) [inline facade/host.go]: Make the audit log message unique "Adding Host with common key" or something like that

File: facade/host.go [hunk 5]
```diff
@@ -115,6 +130,54 @@ func (f *Facade) addHost(ctx datastore.Context, entity *host.Host) ([]byte, erro
...
+	// verify that there are no virtual IPs with the given host IP(s)
+	for _, ip := range entity.IPs {
+		if exists, err := f.HasIP(ctx, pool.ID, ip.IPAddress); err != nil {
+			return nil, fmt.Errorf("error verifying ip %s exists: %v", ip.IPAddress, err)
+		} else if exists {
+			return nil, fmt.Errorf("pool already has a virtual ip %s", ip.IPAddress)
+		}
+	}
+
+	/*
```
Comment (jplouis) [inline facade/host.go]: remove unused code

File: rpc/master/hosts_client.go [hunk 6]
```diff
@@ -56,6 +56,15 @@ func (c *Client) AddHost(host host.Host) ([]byte, error) {
 	return response, nil
 }
 
+//AddHost adds and registers a host with an agreed-upon shared secret keypair
```
Comment (jplouis) [inline rpc/master/hosts_client.go]: Document return value

File: rpc/rpcutils/authcodec.go [hunk 7]
```diff
@@ -28,7 +28,13 @@ import (
 
 var (
 	// RPC Calls that do not require authentication or who handle authentication separately:
-	NonAuthenticatingCalls = []string{"Master.AuthenticateHost", "Agent.BuildHost", "ControlCenterAgent.Ping"}
+	NonAuthenticatingCalls = []string{
+		"Master.AuthenticateHost",
+		"Agent.BuildHost",
+		"ControlCenterAgent.Ping",
+		"Agent.AddHostPrivate",
```
Comment (jplouis) [inline rpc/rpcutils/authcodec.go]: Why do we need agent and master added to this list?

File: cli/cmd/host.go (hunk 1 above)
Comment (ed-) [reply cli/cmd/host.go]: Done via `os.Exit(1)` in the body

**Pull request merged**